### PR TITLE
Add --prod flag to all npm build steps to fix SRI

### PR DESCRIPTION
**Pull request by chmreid**
This PR modifies package.json to add the `--prod` flag to the build step
for the staging stage (not just for the prod stage).

This is necessary because without the `--prod` flag, the staging site
adds a `<noscript></noscript>` tag block that contains `<link>` tags
in it that do not have an 'integrity' attribute.

In other words, if we do not build our site with the `--prod` flag, it
disables SRI for resources in the `<noscript></noscript>` tags, which
leaves users with Javascript **off** vulnerable to CSS-based cross-site
scripting attacks.

See [SEAB-3225](https://ucsc-cgl.atlassian.net/browse/SEAB-3225).

### CODE REVIEW

**Comment by chmreid** [comment]
CircleCI is failing on unit tests at the build step, but it doesn't seem to be related to the new flag, it's a problem with an unreachable (CircleCI) URL:

```
+ java -jar openapi-generator-cli.jar generate -i https://11390-33383826-gh.circle-artifacts.com/0/tmp/artifacts/swagger.yaml -g typescript-angular -o src/app/shared/swagger -c swagger-config.json --skip-validate-spec
[main] ERROR io.swagger.v3.parser.util.RemoteUrl - unable to read
```

**Review by garyluu — CHANGES_REQUESTED** [review]
1. Seems more like an issue with staging running the wrong command
2. `--prod` is deprecated, change to `-c production`

**Comment by garyluu** [comment]
> CircleCI is failing on unit tests at the build step, but it doesn't seem to be related to the new flag, it's a problem with an unreachable (CircleCI) URL:
> 
> ```
> + java -jar openapi-generator-cli.jar generate -i https://11390-33383826-gh.circle-artifacts.com/0/tmp/artifacts/swagger.yaml -g typescript-angular -o src/app/shared/swagger -c swagger-config.json --skip-validate-spec
> [main] ERROR io.swagger.v3.parser.util.RemoteUrl - unable to read
> 
> ...
> 
> [main] ERROR io.swagger.v3.parser.util.RemoteUrl - unable to read
> java.io.FileNotFoundException: https://11390-33383826-gh.circle-artifacts.com/0/tmp/artifacts/swagger.yaml
> 
> ...
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! dockstore-ui2@2.3.0 prebuild: `ts-node -O '{"module":"commonjs"}' git.version.ts && ./scripts/generate-openapi-script.sh`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the dockstore-ui2@2.3.0 prebuild script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /home/circleci/.npm/_logs/2021-08-10T17_17_03_390Z-debug.log
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! dockstore-ui2@2.3.0 prebuild.prod: `npm run prebuild`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the dockstore-ui2@2.3.0 prebuild.prod script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /home/circleci/.npm/_logs/2021-08-10T17_17_03_442Z-debug.log
> 
> 
> Exited with code exit status 1
> ```

Looks like your base branch is probably from a very long time ago

**Comment by chmreid** [comment]
Indeed I had created the branch from an old commit on develop.

**Review by garyluu — APPROVED** [review]
Angular actually defaulted to prod build on newly generated projects so this is fine (to always prod build).

**Review by denis-yuen — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -14,8 +14,8 @@
     "webservice": "./scripts/run-webservice-script.sh",
     "prebuild": "ts-node -O '{\"module\":\"commonjs\"}' git.version.ts && ./scripts/generate-openapi-script.sh",
     "prebuild.prod": "npm run prebuild",
-    "build.prod": "npx ng build --prod --subresource-integrity",
-    "build": "npx ng build --subresource-integrity",
+    "build.prod": "npx ng build -c production --subresource-integrity",
+    "build": "npx ng build -c production --subresource-integrity",
```
Comment (coverbeck) [inline package.json:18]: This is only used for builds running on developers' machines.

Do we want this there? Might be OK, just checking. Presumably it's a slower build?

**Review by coverbeck — COMMENTED** [review]
I'm a bit confused by this.

1. [This](https://github.com/dockstore/dockstore-ui2/blob/55141eb0be076f14b9f3ba86325be651afbbba93/.circleci/config.yml#L208) is where the UI is built and pushed up to s3. It is invoking `npm run build.prod` which is using the (now deprecated) `--prod` flag.
2. We do not have environment-specific UI builds. A given build can be deployed to dev, staging, or prod. Although only tagged releases are typically deployed to staging/prod, there is no difference in how they are built.

I'm wondering if we understand what is really going on? Does prod exhibit the same behavior as staging? (In the JIRA ticket you say prod is dockstore.net, which it's not, and I'm not sure if you meant dockstore.org, or if you compared dev.dockstore.net to staging). Right now the builds on staging and prod should be identical.

dev.dockstore.net and and the develop branch have a new major version of Angular; is it possible Angular fixed something in this area?

The change seems fine, especially if the old option has been deprecated. But I want to make sure we understand what's going on, e.g., if we cherry-pick this as a hotfix (if deemed necessary), will it address the problem?

File: package.json (hunk 1 above)
Comment (coverbeck) [reply package.json:18]: I saw Gary's comment after I submitted my review:

> Angular actually defaulted to prod build on newly generated projects so this is fine (to always prod build).

So it looks like it's fine.

**Comment by chmreid** [comment]
> dev.dockstore.net and and the develop branch have a new major version of Angular; is it possible Angular fixed something in this area?

Or broke something. On a closer look at prod and dev, I noticed that prod does not have the `<noscript></noscript>` block that's missing the SRI, but dev does. (The missing SRI attribute was reported by Netsparker and is what prompted this pull request.) Staging also does not have the `<noscript></noscript>` block, which is consistent with it being on the same version of Angular as prod.

This all seems consistent with the hypothesis that the new version of Angular is adding in the `<noscript></noscript>` block, but that that block leaves out subresource integrity attributes in the resources it puts in there. Which would mean this pull request is not fixing the right thing.

**Comment by chmreid** [comment]
(And that we've uncovered a possible bug in Angular!)

**Review by coverbeck — APPROVED** [review]
@garyluu, it seems like we should still take this, right, to update usage of the deprecated `--prod`, right?

**Comment by garyluu** [comment]
> @garyluu, it seems like we should still take this, right, to update usage of the deprecated `--prod`, right?

my other PR includes this change so it probably doesn't matter either way

**Comment by garyluu** [comment]
Seems to be "fixed" with:

```
gluu@DESKTOP-TB9EL7G:~/dockstore-ui2$ git diff angular.json
diff --git a/angular.json b/angular.json
index 369d142c..af899e07 100644
--- a/angular.json
+++ b/angular.json
@@ -63,7 +63,12 @@
                   "replace": "src/environments/environment.ts",
                   "with": "src/environments/environment.prod.ts"
                 }
-              ]
+              ],
+             "optimization": {
+                "styles": {
+                  "inlineCritical": false
+                }
+              }
             },
```

not entirely sure why

**Comment by chmreid** [comment]
> Seems to be "fixed" with:

Thanks for figuring that out, Gary. When you say "fixed", what do you mean (and why the quotes)? Is the above change included in a different PR? If not, will you open a PR with that change?

**Comment by garyluu** [comment]
> > Seems to be "fixed" with:
> 
> Thanks for figuring that out, Gary. When you say "fixed", what do you mean (and why the quotes)? Is the above change included in a different PR? If not, will you open a PR with that change?

I say "fixed" because I'm not entirely sure why it fixes it and where the problem lies (did Angular screw up the inline css feature?).  The above change isn't in any of my current PRs. Since it's your issue, I was thinking you may want to go in more depth to figure out why the fix works.

**Comment by chmreid** [comment]
Sounds good, just wanted to clarify. Will open a separate PR that implements that fix.

**Pull request closed without merging**